lib/prot/to1/msg32.c

```diff
@@ -13,79 +13,114 @@
+	}
 
-	/* Start Body/Begin Object "bo" tag */
-	publickey = NULL;
+#if defined(ECDSA256_DA)
+	eat->eat_ph->ph_sig_alg = FDO_CRYPTO_SIG_TYPE_ECSDAp256;
+#else
+	eat->eat_ph->ph_sig_alg = FDO_CRYPTO_SIG_TYPE_ECSDAp384;
+#endif
 
-	if (!sdo_begin_write_signature(&ps->sdow, &sig, publickey)) {
-		LOG(LOG_ERROR, "Failed in writing the signature\n");
+	if (!ps->n4) {
+		LOG(LOG_ERROR, "TO1.ProveToRV: Nonce4 not found\n");
 		goto err;
 	}
 
-	sdow_begin_object(&ps->sdow);
+	// copy nonce4 and GUID into the struct
+	if (0 != memcpy_s(&payloadbasemap.eatnonce, SDO_NONCE_BYTES,
+		ps->n4->bytes, ps->n4->byte_sz)) {
+		LOG(LOG_ERROR, "TO1.ProveToRV: Failed to copy Nonce4\n");
+		goto err;
+	}
+	payloadbasemap.eatueid[0] = 1;
+	if (0 != memcpy_s(&payloadbasemap.eatueid[1], SDO_GUID_BYTES,
+		ps->dev_cred->owner_blk->guid->bytes, ps->dev_cred->owner_blk->guid->byte_sz)) {
+			LOG(LOG_ERROR, "TO1.ProveToRV: Failed to copy GUID\n");
+			goto err;
+	}
+	payloadbasemap.eatpayloads = NULL;
 
-	/* Write the "ai" tag */
-	sdo_write_tag(&ps->sdow, "ai");
-	sdo_app_id_write(&ps->sdow);
+	// Create the payload as CBOR map. Sign the encoded payload.
+	// Then, wrap the encoded payload a a bstr.
+	if (!fdo_eat_write_payloadbasemap(&ps->sdow, &payloadbasemap)) {
+		LOG(LOG_ERROR, "TO1.ProveToRV: Failed to write EATPayloadBaseMap\n");
+		goto err;
+	}
+	size_t payload_length = 0;
+	if (!sdow_encoded_length(&ps->sdow, &payload_length) || payload_length == 0) {
+		LOG(LOG_ERROR, "TO1.ProveToRV: Failed to read EATPayload length\n");
+		goto err;
+	}
+	ps->sdow.b.block_size = payload_length;
+	// Set the encoded payload into buffer
+	encoded_payloadbasemap =
+		sdo_byte_array_alloc_with_byte_array(ps->sdow.b.block, ps->sdow.b.block_size);
+	if (!encoded_payloadbasemap) {
+		LOG(LOG_ERROR, "TO1.ProveToRV: Failed to alloc for encoded EATPayload\n");
+		goto err;
+	}
+	eat->eat_payload = encoded_payloadbasemap;
 
-	/* Write back the same nonce which was received in msg31 */
-	sdo_write_tag(&ps->sdow, "n4");
-	if (!ps->n4) {
-		LOG(LOG_ERROR, "ps->n4 is empty MSG#32\n");
+	// reset the SDOW block to prepare for the next encoding.
+	sdo_block_reset(&ps->sdow.b);
+	ps->sdow.b.block_size = CBOR_BUFFER_LENGTH;
+	if (!sdow_encoder_init(&ps->sdow)) {
+		LOG(LOG_ERROR, "TO1.ProveToRV: Failed to initilize SDOW encoder\n");
 		goto err;
 	}
 
-	/* FIXME: Move to error handling. If TO1 restarts, we will leak memory
-	 */
-	sdo_byte_array_write_chars(&ps->sdow, ps->n4);
-	sdo_byte_array_free(ps->n4);
-	ps->n4 = NULL;
-
-	/* Write the GUID received during DI */
-	sdo_write_tag(&ps->sdow, "g2");
-	sdo_byte_array_write_chars(&ps->sdow, ps->dev_cred->owner_blk->guid);
-	/* TODO: Add support for epk defined in spec 0.8 */
-	sdow_end_object(&ps->sdow);
-
-	/* Fill in the pk and sg based on Device Attestation selected */
-	if (sdo_end_write_signature(&ps->sdow, &sig) != true) {
-		LOG(LOG_ERROR, "Failed in writing the signature\n");
+	// generate the signature on encoded payload
+	if (0 !=
```

Comment:
Does 0 indicate a sign failure? If so then it would be better to define a constant and use it.

Reply:
0 indicates success and -1 indicates failure. The return value description is provided in the comment of 'sdo_device_sign'.